richtext-ui/src/commonMain/kotlin/com/halilibo/richtext/ui/string/Text.kt

```diff
@@ -1,49 +1,135 @@
+import androidx.compose.runtime.invalidateGroupsWithKey
+import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.mutableStateListOf
+import androidx.compose.runtime.mutableStateMapOf
+import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
 import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
 import androidx.compose.ui.text.AnnotatedString
+import androidx.compose.ui.text.SpanStyle
 import androidx.compose.ui.text.TextLayoutResult
 import androidx.compose.ui.text.style.TextOverflow
 import com.halilibo.richtext.ui.ClickableText
 import com.halilibo.richtext.ui.RichTextScope
 import com.halilibo.richtext.ui.currentContentColor
 import com.halilibo.richtext.ui.currentRichTextStyle
 import com.halilibo.richtext.ui.string.RichTextString.Format
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.sample
+import kotlinx.coroutines.launch
+import kotlin.time.Duration.Companion.milliseconds
 
 /**
  * Renders a [RichTextString] as created with [richTextString].
  *
  * @sample com.halilibo.richtext.ui.previews.TextPreview
  */
+@OptIn(FlowPreview::class)
 @Composable
 public fun RichTextScope.Text(
   text: RichTextString,
   modifier: Modifier = Modifier,
   onTextLayout: (TextLayoutResult) -> Unit = {},
   softWrap: Boolean = true,
+  isLeafText: Boolean = true,
+  animate: Boolean = false,
+  textFadeInMs: Int = 600,
+  debounceMs: Int = 150,
   overflow: TextOverflow = TextOverflow.Clip,
   maxLines: Int = Int.MAX_VALUE
 ) {
   val style = currentRichTextStyle.stringStyle
   val contentColor = currentContentColor
+  val coroutineScope = rememberCoroutineScope()
   val annotated = remember(text, style, contentColor) {
     val resolvedStyle = (style ?: RichTextStringStyle.Default).resolveDefaults()
     text.toAnnotatedString(resolvedStyle, contentColor)
   }
+  val animations = remember { mutableStateMapOf<Int, TextAnimation>() }
+  val textToRender = remember { mutableStateOf(AnnotatedString("")) }
+
+  if (animate) {
+    val lastAnimatedIndex = remember { mutableIntStateOf(0) }
+    val readyToAnimateText = remember { mutableStateOf(AnnotatedString("")) }
+    // In case no changes happen for a while, we'll render after some timeout
+    val debouncedTextFlow = remember { MutableStateFlow(AnnotatedString("")) }
+    val debouncedText by remember { debouncedTextFlow.debounce(debounceMs.milliseconds) }
+      .collectAsState(AnnotatedString(""), coroutineScope.coroutineContext)
+
+    LaunchedEffect(text) {
+      debouncedTextFlow.value = annotated
+      // If we detect a new phrase, kick off the animation now.
+      if (annotated.hasNewPhraseFrom(textToRender.value.text)) {
+        readyToAnimateText.value = annotated
+       }
+    }
+    LaunchedEffect(isLeafText) {
```

Comment:
Do you need this?
```suggestion
    LaunchedEffect(isLeafText, annotated) {
```

Reply:
yeah, because it allows us to short-circuit the debounce when we finish rendering an astnode

Reply:
I mean the adding annotated. If that changes, you probably want to update the flow.